### ISSUE TITLE
[Question]: Intended / idiomatic use 

### ISSUE THREAD

**Issue by alimoeeny** [ProcedureKit/ProcedureKit#764]
Hi @danthorpe , and others

**tldr**:
What is the idiomatic / intended way of dealing with branches in a queue or in the logical flow of the app. Like branching off to show a login UI if the users session has expired.

**Longer story**
So imagine a scenario in which you want to call an api get a list of things and display them in a tableview.
Steps involved are like,
- make sure already authenticated, if not send them to login or  register
then
- call the API with proper auth tokens, if failed due to auth send back one step, if failed for other reasons retry
-  parse the data
- refresh the UI

**Question:**
what is the intended use or ProcedureKit in the scenario above, is it one of the two options below or something else:
1- create a global flow manager / coordinator which `observes` a single Auth procedure that depending on the state of the app `produces` a next procedure, which can be a UI for login or a network procedure to call the api. Each of these are added to the queue by the coordinator and are in turn being `observed` by the  coordinator for the next state they `produce` .

2- create an Auth procedure a data retrieval procedure and a data display procedure make them depend on each other. The Auth procedure has a condition which depends on a login procedure (should this be added to the same queue?), the login procedure is a UI procedure but before displaying any UI checks for presence of a cookie , if the cookie is not there then show the ui otherwise just `finish`s.

**Comment by danthorpe** [comment]
Hi @alimoeeny - really good question. :+1:

There isn't really a definitive or idiomatic: _this is how you should do this_. Both options you describe above would work, and are possible ways to go.

I would probably create or utilise separate `Procedure` instances for each action (using result injection):

1. check authentication?
2. perform network request - and do some error mapping
3. parse the data
4. refresh the UI
5. present user with authentication dialog.

Then, I could create a `GroupProcedure` subclass, and possibly a custom `Condition`, essentially you want to do this:

### Group
Your group procedure does the happy path of download -> parse. It's input types are the resource you to download, and it's output types are the result of your parsing.

### UI
The update UI procedure will depend on the group, and has an input type of the parsed data. 

### Authentication
You could handle authentication as a condition on the Group - where you'll ensure the user is authorised before running the group. Alternatively, you could respond to failed authentication as an observer of the network procedure, either inside the `GroupProcedure` itself - where you could inspect the error, and then add a `AuthorizeProcedure` to the group which is a dependency on another network procedure.

---

There are a couple of ideas like `RetryProcedure` and `CapabilityProtocol` in the framework to help with this sort of thing. But, they're a little bit advanced - and so might be better for you to create a custom `GroupProcedure` - and deal with the state machine of authentication -> network inside the group.

**Comment by alimoeeny** [comment]
Really appreciate your detailed answer @danthorpe 
I didn't realize, now I do, that you can add procedures to the queue or add dependencies at run time, which was silly of me. 
Thanks a lot and keep up the great work. 

On another note, if you create a todo list or help wanted list for people who want to contribute just a small bit at a time, it may help. I mean sometimes I have a short period of free time that I would love to help with great projects like ProcedureKit and having a list of small bits would facilitate that kinda of contribution.
